rotkehlchen/chain/ethereum/modules/balancer/utils.py

```diff
@@ -0,0 +1,284 @@
+        token1=token1,
+        amount0_in=amount0_in,
+        amount1_in=AssetAmount(ZERO),
+        amount0_out=AssetAmount(ZERO),
+        amount1_out=amount1_out,
+    )
+    return amm_swap
+
+
+def deserialize_token_price(
+        raw_token_price: Dict[str, Any],
+) -> Tuple[ChecksumEthAddress, Price]:
+    """May raise DeserializationError"""
+    try:
+        token_address = raw_token_price['id']
+        usd_price = deserialize_price(raw_token_price['price'])
+    except KeyError as e:
+        raise DeserializationError(f'Missing key: {str(e)}.') from e
+
+    try:
+        token_address = to_checksum_address(token_address)
+    except ValueError as e:
+        raise DeserializationError(
+            f'Invalid ethereum address: {token_address} in token price.',
+        ) from e
+
+    return token_address, usd_price
+
+
+def deserialize_token_day_data(
+        raw_token_day_data: Dict[str, Any],
+) -> Tuple[ChecksumEthAddress, Price]:
+    """May raise DeserializationError"""
+    try:
+        token_address = raw_token_day_data['token']['id']
+        usd_price = deserialize_price(raw_token_day_data['priceUSD'])
+    except KeyError as e:
+        raise DeserializationError(f'Missing key: {str(e)}.') from e
+
+    try:
+        token_address = to_checksum_address(token_address)
+    except ValueError as e:
+        raise DeserializationError(
+            f'Invalid ethereum address: {token_address} in token day data.',
+        ) from e
+
+    return token_address, usd_price
+
+
+def calculate_amm_trade_from_amm_swaps(swaps: List[AMMSwap]) -> AMMTrade:
+    """AMMTrade is always a buy.
+
+    [USDC -> AMPL]               BASE_QUOTE pair is AMPL_USDC.
+    [USDC -> AMPL, AMPL -> WETH] BASE_QUOTE pair is WETH_USDC.
+    """
+    amm_trade = AMMTrade(
+        trade_type=TradeType.BUY,
+        base_asset=swaps[-1].token1,
+        quote_asset=swaps[0].token0,
+        amount=swaps[-1].amount1_out,
+        rate=Price(swaps[-1].amount1_out / swaps[0].amount0_in),
+        swaps=swaps,
+        trade_index=0,
+    )
+    return amm_trade
+
+
+def get_trades_from_tx_swaps(
+        swaps: List[AMMSwap],
+        is_close_amount_mode: bool = False,
+) -> List[AMMTrade]:
+    """Aggregates N AMMSwaps in an AMMTrade.
+
+    We expect that the swaps likely to be aggregated in a single trade are in
+    sequence (the next). This function currently does not has interleaved
+    matching capabilities.
+
+    In the Balancer protocol, when swaps happen within the same pool, the
+    `amount1_out` of the swap N matches the `amount0_in` of the swap N+1.
+    However when swaps happend between different pools these quantities may
+    have a tiny difference (~10e-11). By default two swaps with these tiny
+    differences between quantities would count as two trades. When the
+    `is_close_amount_mode` is enabled we match them in a single trade.
+
+    TODO: `is_close_mode` is experimental. It would be useful to gather more
```

Comment:
As discussed in the standup that sounds way too experimental. Please talk with the balancer people and figure out why this happens.